### PR TITLE
runtime: Enable customization of parallel workers

### PR DESCRIPTION
**Pull request by eman2673**
https://github.com/cucumber/cucumber-js/issues/1044
Tried to cover all the bases but I'm not very familiar with them. However, I'm convinced :)! Awaiting some feedback. I hope something like this can make it into the framework.

### CODE REVIEW

**Comment by eman2673** [comment]
### Updated Read Me Contents

Parallelization is achieved by having multiple child processes running scenarios.

#### Customizable work assignment
Cucumber exposes customization of worker assignment via `setParallelCanAssign`.
The example below overrides the default, `() => true` which processes test cases indiscriminately, with a 1, 2 skip a few processing scheme. This means, the first
2 *remaining* steps will be executed then the 4th *remaining* step will be executed.

```typescript
import { setParallelCanAssign } from '@cucumber/cucumber'
const counter = 0;
setParallelCanAssign((pickleInQuestion, picklesInProgress) => counter++ % 5 < 1)
/** Processing order of: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
 * After   Worked | Waiting    
 * first 2 [0, 1] | [2, 3, 4, 5, 6, 7, 8, 9]
 * 4th     [0, 1, 5] | [2, 3, 4, 6, 7, 8, 9]
 * first 2 [0, 1, 5, 2, 3] | [4, 6, 7, 8, 9]
 * 4th     [0, 1, 5, 2, 3, 8] | [4, 6, 7, 9]
 * fisrt 2 [0, 1, 5, 2, 3, 8, 4, 6] | [7, 9]
 * 4th     [0, 1, 5, 2, 3, 8, 4, 6, 9] | [7]
 * first   [0, 1, 5, 2, 3, 8, 4, 6, 9, 7]
 */ 
```

As you can see, the coordinator always returns to the beginning of the list when 
assigning work. Therefore, on assignment 3 `[2, 3, 4]` will be skipped, but the 
next assignment will check test case `2` as it is the first unworked test case in 
the list.

It is also important to note that this processing scheme could potentially result in 
all workers becoming idle. However, Cucumber prevents this by assigning work regardless 
of the custom handler if no work is in progress. This example is purely explanatory as 
there is no benefit to randomly skipping tests. Workers become idle after checking all 
remaining test cases against the handler. Assignment is attempted on all idle workers 
when a busy worker becomes `ready`.

#### Coordinator
- load all features, generate test cases
- broadcast `test-run-started`
- create workers and for each worker
  - send an `initialize` command
  - when a worker outputs a `ready` command, send it a `run` command with a test case. If there are no more test cases, send a `finalize` command
  - when a worker outputs a `ready` command
    - if there are no more test cases, send a `finalize` command
    - identify the next processable test case (the next test by default)
    - when there are no processable test cases all idle workers remain idle
    - send a `run` command with the test case to an idle worker 
    - repeat if there are still idle workers
  - if all workers become idle then exit in failure
  - when a worker outputs an `event` command,
    broadcast the event to the formatters,
    and on `test-case-finished` update the overall result

File: src/runtime/parallel/README.md [hunk 1]
```diff
@@ -1,11 +1,51 @@
...
 - create workers and for each worker
   - send an `initialize` command
-  - when a worker outputs a `ready` command, send it a `run` command with a test case. If there are no more test cases, send a `finalize` command
+  - when a worker outputs a `ready` command
+    - if there are no more test cases, send a `finalize` command
+    - identify the next processable test case (the next test by default)
+    - when there are no processable test cases all idle workers remain idle
+    - send a `run` command with the test case to an idle worker 
+    - repeat if there are still idle workers
+  - if all workers become idle then exit in failure
```
Comment (charlierudolph) [inline src/runtime/parallel/README.md:48]: The line `if all workers become idle then exit in failure` seems to contrast with the earlier comment:

```
It is also important to note that this processing scheme could potentially result in 
all workers becoming idle. However, Cucumber prevents this by assigning work regardless 
of the custom handler if no work is in progress
```

I think the idea of exiting with failure if the `setParalllelCanAssign` returns false such that all workers end up idle as that means something is wrong with the users code and I'd rather alert them about it then ignoring it.

File: src/runtime/parallel/README.md [hunk 2]
```diff
@@ -1,11 +1,51 @@
...
+ * After   Worked | Waiting    
+ * first 2 [0, 1] | [2, 3, 4, 5, 6, 7, 8, 9]
+ * 4th     [0, 1, 5] | [2, 3, 4, 6, 7, 8, 9]
+ * first 2 [0, 1, 5, 2, 3] | [4, 6, 7, 8, 9]
+ * 4th     [0, 1, 5, 2, 3, 8] | [4, 6, 7, 9]
+ * fisrt 2 [0, 1, 5, 2, 3, 8, 4, 6] | [7, 9]
+ * 4th     [0, 1, 5, 2, 3, 8, 4, 6, 9] | [7]
+ * first    [0, 1, 5, 2, 3, 8, 4, 6, 9, 7]
+ */ 
+```
```
Comment (charlierudolph) [inline src/runtime/parallel/README.md:23]: This example seems very complex to me. Thoughts on trying to make the example something like "will not run a pickle with the word 'complex' in the name if there is a running pickle with the word 'complex' in the name

Comment (charlierudolph) [reply src/runtime/parallel/README.md:23]: I think using names or tags in an example would be good as that I how I would anticipate users actually using this.

File: features/parallel_custom_assign.feature [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        Scenario: a
+          Then value is 1
+
+        Scenario: b
+          Then value is 1
+      """
+    When I run cucumber-js with `--parallel 2`
+    Then it passes
+
+  Scenario: assignment is appropriately applied and fails at last processed scenario 'a'
```
Comment (charlierudolph) [inline features/parallel_custom_assign.feature:51]: I don't really understand this test. It seems like the test run would fail with the same error if running tests serially / in parallel without custom work partitioning / in parallel with custom work partitioning

**Comment by charlierudolph** [comment]
Thank you very much for the contribution @eman2673 and helping get this feature off the ground

File: src/runtime/parallel/README.md (hunk 1 above)
Comment (eman2673) [reply src/runtime/parallel/README.md:48]: You're right. I was failing initially then decided not to fail and overlooked this. My preference is to continue processing. Warning is in fact a great idea. I update the read me and see if I can figure out how to properly log that warning.

File: src/runtime/parallel/README.md (hunk 2 above)
Comment (eman2673) [reply src/runtime/parallel/README.md:23]: Sure thing. Will update accordingly.

File: features/parallel_custom_assign.feature (hunk 3 above)
Comment (eman2673) [reply features/parallel_custom_assign.feature:51]: I may not fully understand how the provided steps work. As I consider your comments, you are probably right. No matter what, the cause of failure will always be the same. My intentions were to validate that Scenario 1 was processed last.

**Comment by eman2673** [comment]
@charlierudolph, thanks for the feedback. I have addressed your concerns. Regarding the warning upon all idle workers, I used `console.warn`. I'm not sure that's the way to go and if it's not just let me know. I'm thinking the alternative is to broadcast the warning to the formatters (I'll have to figure out how to do that).

File: src/runtime/parallel/README.md [hunk 4]
```diff
@@ -1,11 +1,59 @@
...
+
+| Time | WIP | Events |
+|---|---|---|
+| 0 |  | assigned `1 (@complex)` to `worker A` | 
+| 0 | `@complex` | skip `2 & 3 (@complex)` - assign `4 (@simple)` to `worker B` |
+| 2 | `@complex` | skip `2 & 3 (@complex)` - assign `5 (@simple)` to `worker B` |
+| 3 | `@simple` | assign `2 (@complex)` to `worker A` |
+| 4 | `@complex` | skip `3 (@complex)` - assign `6 (@simple)` to `worker B` |
+| 6 |  | assign `3 (@complex)` to `worker A` |
+| 9 |  | done |
```
Comment (charlierudolph) [inline src/runtime/parallel/README.md:28]: Thanks for the update! Thoughts on using this example as the feature test? For assertions I think we could parse the messages and validate the test case started / finished times to ensure no complex tests ran at the same time.

File: src/runtime/parallel/coordinator.ts [hunk 5]
```diff
@@ -213,18 +250,67 @@ export default class Coordinator {
     _.times(numberOfWorkers, (id) =>
       this.startWorker(id.toString(), numberOfWorkers)
     )
-    this.onFinish = done
+    this.onFinish = (status) => {
+      if (this.idleInterventions > 0) {
+        console.warn(
+          `WARNING: All workers went idle ${this.idleInterventions} time(s). Consider revising handler passed to setParallelCanAssign.`
+        )
```
Comment (charlierudolph) [inline src/runtime/parallel/coordinator.ts:257]: Nice, I like this idea. I think using `console.warn` is sufficient. No need to send this to the formatters

Comment (charlierudolph) [reply src/runtime/parallel/coordinator.ts:257]: Can we add a feature tests to show this idleIntervention logic works as expected? (I think the current first feature test would print this)

File: src/runtime/parallel/coordinator.ts [hunk 6]
```diff
@@ -141,6 +160,24 @@ export default class Coordinator {
     }
   }
 
+  awakenWorkers(): void {
+    const workers = _.values(this.workers)
+    _.each(workers, (worker): boolean => {
```
Comment (charlierudolph) [inline src/runtime/parallel/coordinator.ts:165]: I think you can avoid using `_.values` and just pass `this.workers` directly into `_.each`

File: src/runtime/parallel/coordinator.ts [hunk 7]
```diff
@@ -47,12 +60,14 @@ export default class Coordinator {
   private nextPickleIdIndex: number
```
Comment (charlierudolph) [inline src/runtime/parallel/coordinator.ts:60]: Given these changes, I'm not sure the use of pickleIds and nextPickleIdIndex really makes sense anymore. Thoughts on switching making pickleIds to be editable and we splice out items as they are assigned out? We can make a copy of pickleIds and name it something like pickleIdsToExecuted if we need to retain the original list.

File: src/runtime/parallel/coordinator.ts (hunk 5 above)
Comment (eman2673) [reply src/runtime/parallel/coordinator.ts:252]: Having trouble figuring out how to test if something printed to `std.out`. You are absolutely right though, the first test prints this warning, so we just need to add onto that scenario. Once we figure out what that addition needs to be.

File: src/runtime/parallel/coordinator.ts (hunk 7 above)
Comment (eman2673) [reply src/runtime/parallel/coordinator.ts]: I updated the code refactoring out `nextPickleIdIndex`. Looks a bit cleaner to me. I made `pickleIds` a copy of the constructor option to be safe. I don't see a need to host both the original and a copy in the coordinator at this time.

File: src/runtime/parallel/README.md (hunk 4 above)
Comment (eman2673) [reply src/runtime/parallel/README.md:28]: Done! I totally coded to avoid parsing though. Flagged the workers A and B based on the first case they pick up. Then expected A to work complex in order (1-3) and B to work simple in order (1-3). There was a catch though with the third complex test. Both workers are ready when it's time to assign that test so there is no telling which one will pick it up, I took that into consideration.

**Comment by eman2673** [comment]
I may have to do some parsing after all. Locally, I get consistent test results but it looks like I can't depend on the server giving me 2 equally responsive workers. This [run](https://github.com/cucumber/cucumber-js/pull/1588/checks?check_run_id=2082931556) looks like 1 worker finished all 3 scenarios before the other one was ever ready.

**Comment by eman2673** [comment]
I used eval, which I'm not opposed to for testing. If it's a problem then I can make the step_def static and call it a day.

File: src/runtime/parallel/coordinator.ts (hunk 5 above)
Comment (eman2673) [reply src/runtime/parallel/coordinator.ts:251]: @charlierudolph after a bit of a hiatus I finally got back to this and found the `spawn` tag which made it simple.

**Comment by eman2673** [comment]
Thought I was done-ish then `yarn audit` failed. Tired merging from master and that seems to cause more problems. Will try merging from master again in a couple of days and see how that goes.

**Comment by charlierudolph** [comment]
Don't worry about yarn audit as that is unfortunately failing on master. Personally think we need to handle auditing deps differently as not something we can currently deal with (its in a dependency of our dependency and we need to wait for other before we can update).

**Comment by eman2673** [comment]
@charlierudolph, we are good now then. Please let me know if there is anything else we need from this PR.

File: features/step_definitions/parallel_steps.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  handlers: Record<string, Function>
+): void {
+  _.each(envelopes, (envelope) =>
+    _.forOwn(envelope, (value, key) => {
+      ;(handlers[key] || _.noop)(value)
+    })
+  )
+}
+
+function verifyTandem(
```
Comment (charlierudolph) [inline features/step_definitions/parallel_steps.ts:43]: This feels a bit complex to me. Using the handlers and verifying as it parses each envelope while efficient, is a bit hard to follow.

Since our assertion are all on the pickle level, thoughts on having a function that takes in a list of envelopes and returns an array where each element is a list of pickles that are running simultaneously.

Then we can assert that list is empty, or certain things about each list of pickles. I think then could have something like 

```
Then the following pickles run simultaneously:
  | complex1, simple4 |
  | complex1, simple5 |
  | complex2, simple5 |
  | complex2, simple6 |
```
 (To make that more reliable, may need to change time for each (maybe make complex take 350 milliseconds)

File: features/parallel_custom_assign.feature [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+
+        Scenario: b
+          Given scenario 2
+
+        Scenario: c
+          Given scenario 3
+      """
+    When I run cucumber-js with `--parallel 2`
+    Then it passes
+    And it runs tests in order b, c, a
```
Comment (charlierudolph) [inline features/parallel_custom_assign.feature:77]: I'm not certain I see any value in this test that isn't covered by the next test. And since the setParallelCanAssign is less a real world example, think it could just be deleted

File: features/parallel_custom_assign.feature [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+          if (processed !== 3) {
+            expect(worker).to.eq(name)
+            expect(scenario).to.eq(++processed)
+          } else {
+            // Scenario 3 might get picked up by B as both are ready at this point
+            expect(scenario).to.eq(processed++)
+          }
+          setTimeout(cb, delay)
+        }
+      }
```
Comment (charlierudolph) [inline features/parallel_custom_assign.feature:102]: I think if we switch to verifying what runs in parallel, we could delete this step_def code and just have steps set timeout and that's call

File: features/parallel_custom_assign.feature [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+      Given(/^scenario complex (\d+)$/, step_def('A', 300))
+      Given(/^scenario simple (\d+)$/, step_def('B', 200))
+      """
+    And a file named "features/a.feature" with:
+      """
+      Feature: adheres to setParallelCanAssign handler
+        @complex
+        Scenario: 1
```
Comment (charlierudolph) [inline features/parallel_custom_assign.feature:111]: Can we name the scenarios complex# and simple# to make it easier for us in assertions?

File: features/parallel_custom_assign.feature [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+            expect(scenario).to.eq(++processed)
+          } else {
+            // Scenario 3 might get picked up by B as both are ready at this point
+            expect(scenario).to.eq(processed++)
+          }
+          setTimeout(cb, delay)
+        }
+      }
+
+      Given(/^scenario complex (\d+)$/, step_def('A', 300))
```
Comment (charlierudolph) [inline features/parallel_custom_assign.feature:104]: Thoughts on these just being
```
Given('complex step', ...)
Given('simple step', ...)

**Comment by charlierudolph** [comment]
Sorry for delayed review. Took me awhile to find time to get back to this

File: features/step_definitions/parallel_steps.ts (hunk 8 above)
Comment (eman2673) [reply features/step_definitions/parallel_steps.ts:43]: I don't see a way around most of what I'm doing. Given a list of envelops there seems to be no easy way to determine the simultaneously running pickles. From what I gather, to derive this information we must follow a chain that dives past the pickle level.
```
Pickle
pickleId -> testCase
testCaseId -> testCaseStarted
testCaseStartedId -> testCaseFinished
```
I'm using the `started`  & `finished` states to determine parallel processing. I have yet to discover another option. Granted, I'm not too familiar with cucumberjs inner workings. 

Also, while I agree that simplicity is a necessity I'm not sure it wins out in this case if the test will become less reliable and slower.  

If there is no other way to do this then I can attempt to refactor it to be more clear (maybe through verbosity). Please provide additional direction.

Thanks!

File: features/parallel_custom_assign.feature (hunk 9 above)
Comment (eman2673) [reply features/parallel_custom_assign.feature:77]: Agreed! Will remove it.

File: features/parallel_custom_assign.feature (hunk 10 above)
Comment (eman2673) [reply features/parallel_custom_assign.feature:102]: We don't actually need to do any validation here. The last step of the scenario will verify everything we need to know. Therefore, just setting the timeout is adequate.

File: features/parallel_custom_assign.feature (hunk 11 above)
Comment (eman2673) [reply features/parallel_custom_assign.feature:111]: Removed the assertions in `step_def` method. At this point there is no added benefit but I'll keep this thought in mind if something changes.

**Comment by eman2673** [comment]
Yay! Workflow probation :)

File: features/step_definitions/parallel_steps.ts (hunk 8 above)
Comment (charlierudolph) [reply features/step_definitions/parallel_steps.ts:43]: I agree you have to dive past the pickle level to actually verify it, I'm just suggesting a different interface so that someone looking at the test can understand what is being verified without looking at the code. 

I don't think things will get slower as we still go through all envelopes once and the extra assertion time to me would be negligible.

Let me try to implement my thoughts.

**Comment by eman2673** [comment]
This kind of makes me sad 😔

**Comment by mattwynne** [comment]
@eman2673 me too!

Could you help me by resolving the merge conflicts, and I'll see if we get get this some attention?

@davidjgoss have you had a chance to look at this at all?

**Comment by davidjgoss** [comment]
@mattwynne this hasn't been on my radar TBH, but I'll take a look this week!

**Comment by eman2673** [comment]
Sure thing. Will try to get to it by tomorrow.

**Comment by davidjgoss** [comment]
@eman2673 looks like something has gone a bit wrong with the merge from `main` - there are still some conflicts and the diff is huge (220 files). I'm going to see if I can resolve it locally.

**Comment by eman2673** [comment]
> @eman2673 looks like something has gone a bit wrong with the merge from `main` - there are still some conflicts and the diff is huge (220 files). I'm going to see if I can resolve it locally.
> 
> Update: that seems to have done it! I'll get back to you soon with some review feedback.

Yeah, I noticed. After resolving all the conflicts, I didn't have it in me to figure out what was going on.

Thanks for taking care of it.

File: features/step_definitions/parallel_steps.ts (hunk 8 above)
Comment (davidjgoss) [reply features/step_definitions/parallel_steps.ts:42]: @eman2673 if it's okay I'm going to experiment a bit with the idea of Charlie's above and see where it leads, it should help me understand the code a bit better as well.

File: src/runtime/parallel/coordinator.ts [hunk 13]
```diff
@@ -94,6 +111,26 @@ export default class Coordinator implements IRuntime {
...
 
+  awakenWorkers(triggeringWorker: IWorker): void {
+    Object.values(this.workers).forEach((worker) => {
+      if (worker.state === WorkerState.idle) {
+        this.giveWork(worker)
+      }
+      return worker.state !== WorkerState.idle
+    })
+
+    let wip: Boolean = false
```
Comment (davidjgoss) [inline src/runtime/parallel/coordinator.ts:122]: Could these few lines not effectively be replaced with:

```ts
const wip = this.inProgressPickles.length > 0
```

Or am I missing something?

Comment (eman2673) [reply src/runtime/parallel/coordinator.ts:122]: `inProgressPickles` is an object. Having to replace the lodash implementation, I found this is the most performant solution.

**Comment by charlierudolph** [comment]
Sorry been so out of touch here. Got some free time and implemented the feature changes I was thinking. As part of it just created a helper that we could export for users to give them a means to get started with this.

Please review @eman2673 @davidjgoss

File: features/parallel_custom_assign.feature [hunk 14]
```diff
@@ -4,99 +4,70 @@ Feature: Running scenarios in parallel with custom assignment
...
+          Given slow step
       """
     When I run cucumber-js with `--parallel 2`
     Then the error output contains the text:
     """
     WARNING: All workers went idle 2 time(s). Consider revising handler passed to setParallelCanAssign.
     """
-    And no tests ran in tandem
-
-  Scenario: Both works run tests when a valid assignment helper is used
```
Comment (charlierudolph) [inline features/parallel_custom_assign.feature]: Didn't think this feature was needed given its just using the default the parallelCanAssign function (ie just standard parallel functionality)

File: src/runtime/parallel/coordinator.ts [hunk 15]
```diff
@@ -224,7 +224,7 @@ export default class Coordinator implements IRuntime {
       supportCodeLibrary: this.supportCodeLibrary,
     })
     return await new Promise<boolean>((resolve) => {
-      for (let i = 0; i <= this.numberOfWorkers; i++) {
+      for (let i = 0; i < this.numberOfWorkers; i++) {
```
Comment (charlierudolph) [inline src/runtime/parallel/coordinator.ts:227]: Without this change, we were actually starting up one more worker than the user wanter

ie if numberOfWorkers is 2, this loop was executing 3 times, i = 0,1,2

**Comment by mattwynne** [comment]
Nice to have you back @charlierudolph! 😁

**Review by davidjgoss — APPROVED** [review]

**Review by davidjgoss — COMMENTED** [review]
Looks like there might be some flakiness in the tests here e.g. https://github.com/cucumber/cucumber-js/runs/5244903679?check_suite_focus=true - going to run a few more times and try to see what's going on.

**Comment by charlierudolph** [comment]
Increased the time to give us a larger buffer which I hope will decrease chance of flakes (at the cost of the test running a bit longer)

**Comment by aslakhellesoy** [comment]
Hi @eman2673,

Thanks for your making your first contribution to Cucumber, and welcome to the Cucumber committers team! You can now push directly to this repo and all other repos under the cucumber organization! 🍾

In return for this generous offer we hope you will:

* ✅ Continue to use branches and pull requests. When someone on the core team approves a pull request (yours or someone else's), you're welcome to merge it yourself.
* 💚 Commit to setting a good example by following and upholding our [code of conduct](https://cucumber.io/conduct) in your interactions with other collaborators and users.
* 💬 Join the [community Slack channel](http://cucumber.io/support#slack) to meet the rest of the team and make yourself at home.
* ℹ️ Don't feel obliged to help, just do what you can if you have the time and the energy.
* 🙋 Ask if you need anything. We're looking for feedback about how to make the project more welcoming, so please tell us!

On behalf of the Cucumber core team,
Aslak Hellesøy
Creator of Cucumber

**Comment by eman2673** [comment]
@charlierudolph @davidjgoss @mattwynne 

Thanks for all the help getting these changes.

**Comment by mattwynne** [comment]
🎉 you're welcome @eman2673, thanks for all your hard work and patience!

If you're interested in doing any more work on Cucumber just let us know, we can always use more hands on the code. As Aslak said, you can chat with us in our [community Slack](https://cucumber.io/community#slack) in the [#committers](https://cucumberbdd.slack.com/archives/C62D0FK0E) channel 😄 

I'm also trying to make myself available for live pairing ⚡ you can book a slot with me [here](https://calendly.com/mattwynne).
